metabase/metabase
# Fix model metadata propagation

This PR mainly fixes two things.

**Metadata propagation**

Customized model metadata was not propagated to some places (e.g., to metadata popovers you see when hovering fields).

When you have a `FieldDimension` instance, you can generally do `dimension.field()` to retrieve the underlying `Field` instances with all the necessary metadata. [The `field` method](https://github.com/metabase/metabase/blob/b0b8bc80526df8dc126c58181b81f034a95e020d/frontend/src/metabase-lib/lib/Dimension.ts#L576) is pretty tricky as it's not always easy to retrieve the field: it can be found [inside `Metadata` instance](https://github.com/metabase/metabase/blob/b0b8bc80526df8dc126c58181b81f034a95e020d/frontend/src/metabase-lib/lib/Dimension.ts#L578) or in [question's results metadata](https://github.com/metabase/metabase/blob/b0b8bc80526df8dc126c58181b81f034a95e020d/frontend/src/metabase-lib/lib/Dimension.ts#L589) or [restored in a minimal form from the pieces of field info that are available](https://github.com/metabase/metabase/blob/b0b8bc80526df8dc126c58181b81f034a95e020d/frontend/src/metabase-lib/lib/Dimension.ts#L644).

If field metadata is present inside the `Metadata` instance, it won't care about the question's results metadata and return the field. For models where results metadata always matters, it has to be merged in the final field metadata. So this PR refactors `FieldDimension.prototype.field` to support that

**Out-of-sync metadata**

You could get incorrect metadata when entering and leaving the model editor. The model page (`/model/:id`) performs a model nested query so that you can build a new question based on the model. The editor should use the original model query. Results metadata for these two kinds of queries can be different (e.g., original model query can have an expression, and it'd be referenced as `["expression", "my_formula"]`. But in the nested context, it'd be referenced using a field literal `["field", "my_formula", { ... }]`

This was solved by just rerunning the query once you enter/leave the editor. `TableInteractive's` column widths reset event was messing that up. So I had to store `previousQueryBuilderMode` to detect the moment a user leaves the editor to make it ignore this event.

### To Verify

1. Create a native model and open its metadata editor
2. Edit some metadata and click "Save changes."
3. Hover the column names and make sure the metadata popover displays the most relevant info (display names, descriptions, semantic types)
4. Refresh the page, make sure it still works fine

### Demo

**Before**

https://user-images.githubusercontent.com/17258145/152594966-506fde8f-9b0f-4d58-960c-a83d7a71b8d0.mp4

**After**

https://user-images.githubusercontent.com/17258145/152594733-7c0c8af6-bec8-423c-ba48-71276fd59e5c.mp4
